credentials/tls/certprovider/provider.go

```diff
@@ -94,6 +95,10 @@ type KeyMaterial struct {
 	Certs []tls.Certificate
 	// Roots contains the set of trusted roots to validate the peer's identity.
 	Roots *x509.CertPool
+	// SPIFFEBundleMap is an in-memory representation of a spiffe trust bundle
+	// map. If this value exists, it will be used to find the roots for a given
+	// trust domain rather than the Roots in this struct.
+	SPIFFEBundleMap spiffe.BundleMap
```

Comment:
This type `spiffe.BundleMap`, is defined in an internal package. But the `KeyMaterial` struct is exported to our users. This will be unfortunate if we add a new field in a publicly exported type, but our users can't have access to that type.

Reply:
Would it be possible instead of the `pemfile` certificate provider to convert the SPIFFE bundle into an `x509.CertPool` and return it using the `Roots` field instead? I see that there is a way to extract the certificates from a SPIFFE bundle. So, you might be able to extract them and create an `x509.CertPool` and add those certificates to that pool and return it. But I don't know if functionally, that will still do what you want it to do.

Reply:
I think the right thing to do here would be to make `spiffe.BundleMap` publicly exported

Tne SPIFFE bundle map can't be directly set as x509.CertPool roots. It's essentially a map of string->root certs, where the string is a SPIFFE ID, and at handshake time you lookup the roots for the peer's SPIFFE ID, and _after_ this map lookup we set the roots. But the `KeyMaterial` call doesn't necessarily happen during the handshake.

We could make it a map of strings to x509.CertPool (which is essentially what it is), but I think it would be okay to publicly export the type - would you rather see a public module that has the `spiffe.BundleMap`, or I think we can typealias and say `pemfile.spiffeBundleMap = spiffe.BundleMap`?

Reply:
I think it would be easaier/nicer if we make `SPIFFEBundleMap` to be of type `map[string]*spiffebundle.Bundle` here and get rid of the `BundleMap` type in the internal `spiffe` package.

We also need to document the `Roots` field that it has lower priority over the `SPIFFEBundleMap` field.

Also, maybe `SPIFFEBundles` is a better name as opposed to `SPIFFEBundleMap`?

Reply:
Removed the type in preference of `map[string]*spiffebundle.Bundle` and notated the preferences in the `Roots` field

The `SPIFFEBundleMap` name here comes from [the spec](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#5-spiffe-bundle-map), rather than because it is a `Map` in Go - I think we should keep the name.